activerecord/lib/active_record/persistence.rb

```diff
@@ -971,6 +971,7 @@ def reload(options = nil)
       end
 
       @association_cache = fresh_object.instance_variable_get(:@association_cache)
+      @association_cache.each_value { |association| association.instance_variable_set(:@owner, self) }
```

Comment:
I wonder what core team folks think on that but I always feel uncomfortable about using `instance_variable_set` unless we are patching some code we have no access to. Otherwise we are using something that wasn't supposed to be used even within the framework itself and makes me think whether we should add a public method to `association` to be able to set the `owner`. To be fair it might not be a solid argument as we already set and access instance variables on the lines above and below this one 🙃 

Other than that, the PR makes sense to me.

Reply:
Yes, this was exactly my thinking!

I am also not very comfortable using `instance_variable_get/instance_variable_set` but in this particular case I decided to go with it because:
1. This is already done one line above and below just as you said :)
2. I wasn't completely sure my approach would be accepted (dunno, maybe some other reasons I am not aware of) so doing some major change, like trying to introduce some new public method into `association` (which in turn might require changes to the docs, etc, etc) so that it can be used in this place, seemed a bit too premature to me.

Reply:
But if you think it's worth it, let me know, and I will try to introduce my change via a new public method. 
However the lines above and below will still be using `instance_variable_get`, so dunno... Changing those in this PR as well seems like a bit too much for such a small problem/PR.

Reply:
Yeah, let's avoid those instance variable set here. If we need to set the owner outside of that object, we should expose a writter.

Reply:
No problem, updated.